src/test/deploy/functions/validate.spec.ts

```diff
@@ -0,0 +1,119 @@
+import { expect } from "chai";
+import * as fsutils from "../../../fsutils";
+import * as validate from "../../../deploy/functions/validate";
+import * as projectPath from "../../../projectPath";
+import * as FirebaseError from "../../../error";
+import * as sinon from "sinon";
+
+// have to require this because no @types/cjson available
+// tslint:disable-next-line
+const cjson = require("cjson");
+
+describe("validate", () => {
+  describe(".functionsDirectoryExists", () => {
+    it("should not throw error if functions directory is present", () => {
+      const ppath = sinon.stub(projectPath, "resolveProjectPath");
+      ppath.returns("some/path/to/project");
+      const listDir = sinon.stub(fsutils, "dirExistsSync");
+      listDir.returns(true);
+
+      expect(validate.functionsDirectoryExists("cwd", "sourceDirName")).to.not.throw;
+      sinon.restore();
```

Comment:
at this level (the describe block for the method), maybe it would be cleaner to include a `beforeEach` and `afterEach` block for your stubs. The stubs can be defined as the happy or sad path, but then behavior of them can change after too. So, for example:

```typescript
describe("functionsDirectoryExists", () => {
  let ppath: sinon.Stub;
  let listDir: sinon.Stub;

  beforeEach(() => {
    ppath = sinon.stub(projectPath, "resolveProjectPath");
    // optionally, a return value can be chained right here for convenience.
    listDir = sinon.stub(fsutils, "dirExistsSync");
  });

  afterEach(() => {
    sandbox.restore();
  });

  it("should not...", () => {
    ppath.returns("some/path/to/project");
    listDir.returns(true);

    expect(...).to.not.throw;
  });
});
```

This applies to the last group of tests as well 🙂

Reply:
added beforeEach and afterEach to all tests using sinon - I thought it was a nice idea to define a happy path stub and a sad path stub, but once I got to implementing it, I realized it became less readable and more complex, as the behavior would hinge only on the arguments that are passed in (so if they were misspelled or wrong in any other way, the tests would break, making them quite frail). In the end I decided against it.

Reply:
Yeah, keeping the stubs _just_ the stubs is fine. Sometimes, if there's an uncomfortable number of things that have to be stubbed, it's better to create the stubs and set their behavior to be the happy path - meaning if you called the method under test with _any_ input, it would pass. Then, you can test failure cases by just having to change a couple stubs at a time (and they get reset to the happy path for the next test). What you ended up with is great!